truffle/src/main/ruby/core/proc.rb

```diff
@@ -33,18 +41,31 @@ class Proc
 
   def curry(curried_arity = nil)
     if lambda? && curried_arity
+
       if arity >= 0 && curried_arity != arity
-        raise ArgumentError, "Wrong number of arguments (%i for %i)" % [
-          curried_arity,
-          arity
-        ]
+        raise ArgumentError, "Wrong number of arguments (#{curried_arity} for #{arity})"
       end
 
-      if arity < 0 && curried_arity < (-arity - 1)
-        raise ArgumentError, "Wrong number of arguments (%i for %i)" % [
-          curried_arity,
-          -arity - 1
-        ]
+      if arity < 0
+        total_args = 0
+        req_args = 0
+        has_rest = false
+
+        parameters.each do |p|
+          type = p.first
+          req_args += (type == :req ? 1 : 0)
+          total_args += (type != :block ? 1 : 0)
+          has_rest |= type == :rest 
+        end
+      
+        if (curried_arity < req_args) ||
+           (!has_rest && (curried_arity > total_args))
```

Comment:
I'm with @nirvdrum on this. I tend to find implicit grouping harder to read when you come back to things six months later.